### ISSUE TITLE
Performance Issues during failover testing with periodic refresh enabled

### ISSUE THREAD

**Issue by nathanielhsu** [lettuce-io/lettuce-core#2114]
Reaching out about an issue I'm facing with periodic refresh during cache node failovers.

From investigation, it looks like when failover happens, periodic refresh is helping in picking up the failover quickly. However, once the new replica node comes up and is still being restored (takes around 30 minutes), periodic refresh is causing the client to try connect to the new replica node, but the node is refusing connection because it is not ready yet. Because of this, I am seeing spikes in Connected/Disconnected events when periodic refresh happens during the time when a new replica node is still being restored. These spikes are causing performance degradation in my service and causing spikes in timeouts to GET calls when they happen.

Any ideas here on how this can be fixed? Or is this a limitation with Lettuce client currently?

Here is the topology options I am using. As I am working with a relatively large cluster, I had set `dynamicRefreshSources` to false to mitigate impact of periodic refresh on the service.
```
        final ClusterTopologyRefreshOptions topologyOptions = ClusterTopologyRefreshOptions.builder()
                .enableAllAdaptiveRefreshTriggers()
                .enablePeriodicRefresh(Duration.ofMinutes(10))
                .dynamicRefreshSources(false)
                .build();
```

Any help or insight here would be appreciated. Thanks!

**Comment by mp911de** [comment]
Let me rephrase your request to double-check I understand the scenario: When adding a cluster node, the newly added node is being queried for its topology during a refresh and you want to avoid that.

In such a case, disabling `dynamicRefreshSources` is the way to go. Lettuce can't know the target node isn't ready yet. When `dynamicRefreshSources` is enabled, the first pass on topology discovery adds nodes, that were not provided in the connection URI (set of initial nodes) are additionally queried for their topology. When `dynamicRefreshSources` is disabled, Lettuce only uses the initial cluster nodes to query the topology.

**Comment by nathanielhsu** [comment]
Thanks for the response here! Let me rephrase my question a bit here. I think it wasn't super clear.

The specific case I was looking at occurred during node failovers. When a new node comes up after the previous node failed and got removed, there is a period of time where the new node is being filled with data from the other healthy nodes. Whenever periodic refresh was happening during this time while the node was still filling up, I was seeing huge spikes in `ConnectedEvent` and `DisconnectedEvent` from the Lettuce client and corresponding spikes in cache call latency. I believe the node is rejecting the connection as it is not ready yet. 

Any ideas as to why this is happening and how this can be resolved? Thanks!

**Comment by mp911de** [comment]
Thanks for the detail. It would make a lot of sense to capture a debug/trace log (`io.lettuce.core.protocol` category) to see when the disconnect is happening and what command has preceded it. 

Data loading can be discovered by inspecting the `master_repl_offset` on the actual node. Typically, any commands during data loading fail with the `LOADING` error and not a disconnect, so I'm not sure what happens here.

**Comment by yangbodong22011** [comment]
@nathanielhsu The process of redis cluster [scaling](https://redis.io/docs/manual/scaling/#adding-a-new-node) is:
1, add newnode (note: visible by `CLUSTER NODES` command)
2, move slot data to newnode

Because in the first step, the cluster topology refresh (call `CLUSTER NODES`) can already see the newnode, so Lettuce will create a connection to the newnode, resulting in the phenomenon you said: ConnectedEvent and DisconnectedEvent corresponding spikes.

I don't think there is a way to avoid this problem on the client side in the current situation, unless Redis Cluster can do it: add nodes, and make the `CLUSTER NODES` command visible until Ready (the data is full).

**Comment by mp911de** [comment]
Thanks for your feedback. Is there something left to do for us? If not, I'd close the ticket.

**Comment by yangbodong22011** [comment]
@mp911de I think there no thing we can do. but before close it, wait @nathanielhsu feedback？

**Comment by nathanielhsu** [comment]
@mp911de @yangbodong22011 thanks for the response here. I'll take a quick look at this to see if there's a workaround. Is there any way to check if a node is still filling up with data before connecting to it that you know of?

**Comment by yangbodong22011** [comment]
> @mp911de @yangbodong22011 thanks for the response here. I'll take a quick look at this to see if there's a workaround. Is there any way to check if a node is still filling up with data before connecting to it that you know of?

When redis loading data, most commands will return `LOADING Redis is loading the dataset in memory`.

**Comment by yangbodong22011** [comment]
@nathanielhsu If you don't have any other questions, please close the issue by the way.

**Comment by nathanielhsu** [comment]
Only question is would it be possible for lettuce to check for this LOADING status during topology refresh prior to making a connection attempt?

**Comment by yangbodong22011** [comment]
> @yangbodong22011 Only question is would it be possible for lettuce to check for this LOADING status during topology refresh prior to making a connection attempt?

I think no, that is not Lettuce's responsibility, we are only faithfully responsible for the results of `CLUSTER NODES`.

**Comment by nathanielhsu** [comment]
Yeah that makes sense. Alright sounds good I'll be closing this issue then. Thanks for the help here
